packages/plugins/typed-variable-modal/src/Modal.js

```diff
@@ -0,0 +1,372 @@
+/**
+ * @license
+ * Copyright 2020 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @fileoverview Class responsible for creating a Blockly modal.
+ * @author aschmiedt@google.com (Abby Schmiedt)
+ */
+
+
+import * as Blockly from 'blockly/core';
+
+/**
+ * Class responsible for creating a Blockly modal.
+ */
+export class Modal {
+
+  /**
+   * Constructor for creating a Blockly modal.
+   * @param {string} title The title for the modal.
+   * @param {!Blockly.WorkspaceSvg} workspace The workspace to display the modal
+   *     over.
+   */
+  constructor(title, workspace) {
+    /**
+     * The title for the modal.
+     * @type {string}
+     * @private
+     */
+    this.title_ = title;
+
+    /**
+     * The workspace to display the modal over.
+     * @type {!Blockly.WorkspaceSvg}
+     * @protected
+     */
+    this.workspace_ = workspace;
+
+    /**
+     * The last focusable element for the modal.
+     * @type {HTMLElement}
+     * @private
+     */
+    this.lastFocusableEl_ = null;
+
+    /**
+     * The first focusable element for the modal.
+     * @type {HTMLElement}
+     * @private
+     */
+    this.firstFocusableEl_ = null;
+
+    /**
+     * HTML container for the modal.
+     * @type {HTMLDivElement}
+     * @private
+     */
+    this.htmlDiv_ = null;
+
+    /**
+     * Array holding info needed to unbind events.
+     * Used for disposing.
+     * Ex: [[node, name, func], [node, name, func]].
+     * @type {!Array.<Array<?>>}
+     * @private
+     */
+    this.boundEvents_ = [];
+
+    /**
+     * If true close the modal when the user clicks outside the modal.
+     * Otherwise, only close when user hits the 'X' button or escape.
+     * @type {boolean}
+     */
+    this.closeOnClick = true;
```

Comment:
Maybe rename to shouldCloseOnOverlayClick?

Reply:
Also, should we add one fore shouldCloseOnEsc?

Reply:
Added both